chainerx_cc/chainerx/cuda/memory_pool.cc

```diff
@@ -37,66 +39,174 @@ MallocStatus PinnedMemoryAllocator::Malloc(void** ptr, size_t bytesize) {
     CHAINERX_NEVER_REACH();
 }
 
+std::unique_ptr<Chunk> Chunk::Split(size_t bytesize) {
+    CHAINERX_ASSERT(bytesize_ >= bytesize);
+    if (bytesize_ == bytesize) {
+        return nullptr;
+    }
+
+    std::unique_ptr<Chunk> remaining = std::make_unique<Chunk>(ptr(), bytesize, bytesize_ - bytesize);
+    bytesize_ = bytesize;
+
+    if (next_ != nullptr) {
+        remaining->SetNext(next_);
+        remaining->next()->SetPrev(remaining.get());
+    }
+    next_ = remaining.get();
+    remaining->SetPrev(this);
+
+    return remaining;
+}
+
+void Chunk::MergeWithNext() {
+    CHAINERX_ASSERT(next_ != nullptr);
+    bytesize_ += next_->bytesize();
+    if (next_->next() != nullptr) {
+        next_->next()->SetPrev(this);
+    }
+    next_ = next_->next();
+}
+
+// Pushes a chunk into an appropriate free list
+//
+// Not thread-safe
+void MemoryPool::PushIntoFreeList(std::unique_ptr<Chunk> chunk) {
+    FreeList& free_list = free_bins_[chunk->bytesize()];
+    free_list.emplace_back(std::move(chunk));
+}
+
+void MemoryPool::CompactFreeBins(std::map<size_t, FreeList>::iterator it_start, std::map<size_t, FreeList>::iterator it_end) {
+    auto it_start_rev = std::make_reverse_iterator(it_start);
+    it_start = std::find_if(it_start_rev, free_bins_.rend(), [](const auto& p) { return !p.second.empty(); }).base();
+    it_end = std::find_if(it_end, free_bins_.end(), [](const auto& p) { return !p.second.empty(); });
+    free_bins_.erase(it_start, it_end);
+}
+
+// Finds best-fit, or a smallest larger allocation if available
+//
+// Not thread-safe
+std::unique_ptr<Chunk> MemoryPool::PopFromFreeList(size_t allocation_size) {
+    auto it_start = free_bins_.lower_bound(allocation_size);
+    size_t distance{0};
+    for (auto it = it_start; it != free_bins_.end(); ++it, ++distance) {
```

Comment:
This loop could also written with `std::find_if` and `std::distance`. The complexity of `std::distance` is O(N), so it would not increase.